docs: Add instructions on how to setup an O3DE scene to work with `rai_sim`. (#431)
RobotecAI/rai: docs/rai_sim/o3de_setup.md

# How to setup an O3DE scene to work with rai_sim

This page is a tutorial on how to set up your O3DE scene to be compatible with the rai_sim and rai_bench frameworks, i.e. to allow for:

- **spawning** and **despawning** objects
- **tracking their position** in real time

O3DE comes with a ROS2 gem that allows for easy integration of your scene with ROS2. We will use the ROS2 Spawner component and ROS2 Frame component to achieve the desired functionality. The ROS2 Spawner component exposes a service that allows you to spawn and despawn objects in the scene, while the ROS2 Frame component put on an entity will publish its position on the /tf topic.

> ## Requirements

Reviewer: I would add a list of required packages to the requirements list (or at least a link to O3DE docs). There are some simulation-related requirements (ROS 2 packages) and some other tools, such as `ninja-build` (or even `clang` - please keep in mind this tutorial is in RAI namespace, hence targeting users with different backgrounds and knowledge).

Author: I added notes with links to the respective o3de documentation pages about all the dependencies and how to configure o3de.

> - https://github.com/o3de/o3de and https://github.com/o3de/o3de-extras checked out on tag **2409.2**

Reviewer: It would be worth reminding the users that both repositories use LFS.

Author: I added a note about that.

- an o3de project created (instructions: https://www.docs.o3de.org/docs/welcome-guide/create/creating-projects-using-cli/creating-linux/)

## **Step 1. Enable ROS2 gem in your project.**

```bash

> scripts/o3de.sh enable-gem -gn ROS2 -pp /path/to/your-project/

Reviewer: If you create an empty project, it will receive `PhysX` Gem by default. `ROS2` Gem depends on `PhysX5`. This way you will create a project with both `PhysX` and `PhysX5`, which leads to a segfault. Add a call to disable `PhysX` Gem.

Author: That's true, thank you for noticing it. I added a note about it.

```

## **Step 2. Build the Editor and open it**

```bash
cd /path/to/your-project/
cmake -B build/linux -S . -G "Ninja Multi-Config" -DLY_STRIP_DEBUG_SYMBOLS=TRUE -DLY_DISABLE_TEST_MODULES=ON

> cmake --build build/linux --config profile --target Editor -j 22

Reviewer: Did you consider using SDK? Building O3DE from the source takes some time.
I know there were some problems with Ubuntu 24.04 and the SDK, not sure if this is still the case with O3DE 2409.2.

Author: I tested the newest 2409.2 SDK package on Ubuntu 24.04 and tried to set up a project using it, but ran into a few issues like a python package not installing correctly and the Project Manager refusing to export the project, or constant crashing when entering game mode in the Editor when the ROS2 gem is enabled.

Reviewer: Thanks for letting me know.

./build/linux/bin/profile/Editor
```

## **Step 3. Set up your spawnable prefabs**

> - Place your desired object somewhere in the scene

Reviewer: I do not understand this instruction. What is the meaning of the _object_? 

I would rather use the O3DE terminology, such as _asset_, _component_, and _entity_. The user should know where to click and how to click. A tutorial should always give a concrete definition of what is being done. E.g. a tutorial could describe adding a blue cube. Users can change it into a green dragon or a red heart asset, but the tutorial should say "place a cube of size 1 at point 0,0,0" instead of "place your desired object somewhere".

Author: I added a simple instruction on how to create an entity and replaced the word `object` with `entity` whenever talking about o3de entities.

> - On the entity which you want to track, add the ROS2 Frame component and configure it as follows:

Reviewer: Your instructions are great for experienced O3DE users, but might be too difficult to understand for people trying O3DE for the first time:
```suggestion
- Select the entity you want to track in the _entity manager_ and add the `ROS2 Frame component` in the _inspector window_; configure it as follows:
```

![image.png](../imgs/rai_sim/o3de/image.png)

Leave the frame name and joint name empty, set namespace strategy to “Generate from entity name” and check “Publish transform”.

> - Save the object as a prefab

Reviewer: ```suggestion
- Right-click on the entity name in the _entity manager_ to save the object as a prefab
```

## **Step 4. Setup ROS2 Spawner**

- Create an empty entity and add a ROS2 spawner component to it.

![image.png](../imgs/rai_sim/o3de/image%203.png)

![image.png](../imgs/rai_sim/o3de/image%204.png)

- Click the “+” icon inside the ROS2 Spawner component window

![image.png](../imgs/rai_sim/o3de/image%205.png)

- Enter the **name by which you will later refer to the object to spawn it**

![image.png](../imgs/rai_sim/o3de/image%206.png)

- Click on the little folder icon that appeared next to your object’s name

![image.png](../imgs/rai_sim/o3de/image%207.png)

- Navigate to the place where you saved your prefab and choose it here and click “OK”

![image.png](../imgs/rai_sim/o3de/image%208.png)

### **Repeat steps 3. and 4. for each object that you want to make spawnable.**

Now, when you press play in the Editor, you should be able to spawn new objects and see their positions on the /tf topic, like so:

```bash
ros2 service call /spawn_entity gazebo_msgs/srv/SpawnEntity "{name: 'apple', initial_pose: {position:{ x: 0.0, y: 0.0, z: 0.2 }, orientation: { x: 0.0, y: 0.0, z: 0.0, w: 1.0 } } }"
```

```bash
ros2 topic echo /tf
```

```bash

> kdabrowski@robo-pc-019:~$ ros2 topic echo /tf

Reviewer: You might want to remove your username and the localhost name
```suggestion
ros2 topic echo /tf
```

Author: I restructured this section a bit to make it clearer.

transforms:
- header:
    stamp:
      sec: 163
      nanosec: 66984981
    frame_id: apple_1/odom
  child_frame_id: apple_1/
  transform:
    translation:
      x: -2.6995063308277167e-05
      y: -2.1816253138240427e-05
      z: 0.0499994121491909
    rotation:
      x: 2.1038686099927872e-05
      y: -5.1415558118605986e-05
      z: -3.404247763683088e-05
      w: 1.0
---
```

If these commands work for you, your scene is successfully set up and ready to be used with rai_sim. You can now build the binary package and run it with rai_sim.

> ## Step 5. Building the binary package

Reviewer: Isn't building a game launcher in _proifle_ enough? If you switch to _release_, there will be another build of 20-30 minutes depending on the machine.

Author: We are building the monolithic game package here, which for me takes additional time on both _profile_ and _release_ configs (after building the Editor in _profile_ of course), but maybe the user doesn't need the monolithic package and is just interested in running the GameLauncher. So I replaced this step with building a GameLauncher in _profile_ configuration and left building the monolithic as an optional step if the user is interested in having a portable package.

# How to setup an O3DE scene to work with rai_sim

This page is a tutorial on how to set up your O3DE scene to be compatible with the rai_sim and rai_bench frameworks, i.e. to allow for:

- **spawning** and **despawning** entities
- **tracking their position** in real time

O3DE comes with a ROS2 gem that allows for easy integration of your scene with ROS2. We will use the ROS2 Spawner component and ROS2 Frame component to achieve the desired functionality. The ROS2 Spawner component exposes a service that allows you to spawn and despawn entities in the scene, while the ROS2 Frame component put on an entity will publish its position on the /tf topic.

## Requirements

> - Ubuntu operating system - the ROS2 gem in O3DE is only supported on Ubuntu

Reviewer: O3DE is officially supported on Fedora as well
```suggestion
- Linux operating system (recommended platform for ROS2 gem in O3DE is Ubuntu 22.04 or Ubuntu 24.04)
```

- an o3de project created ([instructions](https://www.docs.o3de.org/docs/welcome-guide/create/creating-projects-using-cli/creating-linux/)).

## **Step 1. Enable ROS2 gem in your project.**

The ROS2 gem depends on the `PhysX5` gem, which will conflict with the `PhysX` gem that is enabled by default when you create a project in O3DE. Therefore first make sure to have the `PhysX` gem disabled. To do this, enter the directory of your cloned O3DE repository and run the following command:

```bash
scripts/o3de.sh disable-gem -gn PhysX -pp /path/to/your-project/
```

Now you can enable the ROS2 gem:

```bash
scripts/o3de.sh enable-gem -gn ROS2 -pp /path/to/your-project/
```

## **Step 2. Build the Editor and open it**

```bash
cd /path/to/your-project/
cmake -B build/linux -S . -G "Ninja Multi-Config" -DLY_STRIP_DEBUG_SYMBOLS=TRUE -DLY_DISABLE_TEST_MODULES=ON
cmake --build build/linux --config profile --target Editor -j 22
./build/linux/bin/profile/Editor
```

> 

Reviewer: ```suggestion

> Note: This process builds O3DE from the source, which might take up to two hours depending on your hardware.

```

## **Step 3. Set up your spawnable prefabs**

- Create an entity in the scene that you want to make spawnable. For example, to create a simple entity that interacts with the physical environment, you can:
  - Right-click in the _entity manager_ and click _Create entity_.
  - Select the entity and in the _inspector window_ add _PhysX Dynamic Rigid Body_, _Box Shape_ and _PhysX Shape Collider_ components to it.
- Select the entity you want to track in the _entity manager_ and add the `ROS2 Frame component` in the _inspector window_; configure it as follows:

![image.png](../imgs/rai_sim/o3de/image.png)

> Leave the frame name and joint name empty, set namespace strategy to “Generate from entity name” and check “Publish transform”.

Reviewer: ```suggestion
Leave the frame name and joint name empty, set namespace strategy to **Generate from entity name** and check **Publish transform**.
```

- Right-click on the entity name in the _entity manager_ to save the entity as a prefab

![image.png](../imgs/rai_sim/o3de/image%201.png)

![image.png](../imgs/rai_sim/o3de/image%202.png)

You can now delete your entity from the scene.

## **Step 4. Setup ROS2 Spawner**

> - Create an empty entity and add a ROS2 spawner component to it.

Reviewer: Try to keep the docs more consistent with lowercase and capitals when needed.
```suggestion
- Create an empty entity and add a ROS2 Spawner component to it.
```

![image.png](../imgs/rai_sim/o3de/image%203.png)

![image.png](../imgs/rai_sim/o3de/image%204.png)

- Click the “+” icon inside the ROS2 Spawner component window

![image.png](../imgs/rai_sim/o3de/image%205.png)

- Enter the **name by which you will later refer to the entity to spawn it**

![image.png](../imgs/rai_sim/o3de/image%206.png)

- Click on the little folder icon that appeared next to your entity’s name

![image.png](../imgs/rai_sim/o3de/image%207.png)

- Navigate to the place where you saved your prefab and choose it here and click “OK”

![image.png](../imgs/rai_sim/o3de/image%208.png)

### **Repeat steps 3. and 4. for each entity that you want to make spawnable.**

> Now, when you press play in the Editor, you should be able to spawn new entities by calling the `/spawn_entity` service, like so:

Reviewer: ```suggestion
Now, press the **play** button in the Editor to start the simulation. You should be able to spawn new entities by calling the `/spawn_entity` service, like so:
```
